### ISSUE TITLE
getAvailableClassesInRevision seems to create a memory leak

### ISSUE THREAD

**Issue by motjokk** [opensourceBIM/BIMserver#950]
**Description of the bug**
Bimserver seems to be under a very heavy workload when requesting `getAvailableClassesInRevision` on 
bimserver-1.5.125

This appears to be independent of file size.

![image](https://user-images.githubusercontent.com/33512457/54669959-e4a3ec80-4af2-11e9-9317-a3a44f60bdb8.png)

Bimserver.log

```
2019-03-20 08:19:27,425 INFO [localhost-startStop-1] o.b.ServerInfoManager [ServerInfoManager.java:113] Changing server state to RUNNING
2019-03-20 08:19:27,426 INFO [localhost-startStop-1] o.b.BimServer [BimServer.java:1100] Initializing database dependant logic...
2019-03-20 08:19:27,659 INFO [localhost-startStop-1] o.i.IfcOpenShellEnginePlugin [IfcOpenShellEnginePlugin.java:102] Using /var/www/ROOT/tmp/org.opensourcebim.ifcopenshellplugin/IfcGeomServer-v0.6.0-33cbcc2-linux64
2019-03-20 08:19:28,720 INFO [localhost-startStop-1] o.b.BimServer [BimServer.java:1249] Done initializing database dependant logic
2019-03-20 08:19:29,004 INFO [localhost-startStop-1] o.b.s.w.j.Jsr356Impl [Jsr356Impl.java:104] WebSocket context initialized
2019-03-20 08:20:59,551 INFO [ServiceInterface.getAvailableClassesInRevision] o.b.d.DatabaseSession [DatabaseSession.java:1251] Inefficient getMap
2019-03-20 08:23:15,868 INFO [ServiceInterface.getAvailableClassesInRevision] o.b.d.b.BerkeleyKeyValueStore [BerkeleyKeyValueStore.java:588] reads: 6421858
```

```
Catalina.out
0-Mar-2019 08:19:21.446 INFO [localhost-startStop-1] org.apache.jasper.servlet.TldScanner.scanJars At least one JAR was scanned for TLDs yet contained no TLDs. Enable debug logging for this logger for a complete list of JARs that were scanned but no TLDs were found in them. Skipping unneeded JARs during scanning can improve startup time and JSP compilation time.
Logging to /var/www/ROOT/logs/bimserver.log
20-Mar-2019 08:19:34.560 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deploying web application directory [/var/www/ROOT/logs]
20-Mar-2019 08:19:34.574 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deployment of web application directory [/var/www/ROOT/logs] has finished in [14] ms
20-Mar-2019 08:19:34.574 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deploying web application directory [/var/www/ROOT/tmp]
20-Mar-2019 08:19:34.584 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deployment of web application directory [/var/www/ROOT/tmp] has finished in [10] ms
20-Mar-2019 08:19:34.584 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deploying web application directory [/var/www/ROOT/incoming]
20-Mar-2019 08:19:34.593 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deployment of web application directory [/var/www/ROOT/incoming] has finished in [9] ms
20-Mar-2019 08:19:34.594 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deploying web application directory [/var/www/ROOT/maven]
20-Mar-2019 08:19:34.604 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deployment of web application directory [/var/www/ROOT/maven] has finished in [10] ms
20-Mar-2019 08:19:34.604 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deploying web application directory [/var/www/ROOT/cache]
20-Mar-2019 08:19:34.612 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deployment of web application directory [/var/www/ROOT/cache] has finished in [8] ms
20-Mar-2019 08:19:34.612 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deploying web application directory [/var/www/ROOT/plugins]
20-Mar-2019 08:19:34.620 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deployment of web application directory [/var/www/ROOT/plugins] has finished in [8] ms
20-Mar-2019 08:19:34.620 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deploying web application directory [/var/www/ROOT/emailtemplates]
20-Mar-2019 08:19:34.627 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deployment of web application directory [/var/www/ROOT/emailtemplates] has finished in [7] ms
20-Mar-2019 08:19:34.627 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deploying web application directory [/var/www/ROOT/database]
20-Mar-2019 08:19:34.635 INFO [localhost-startStop-1] org.apache.catalina.startup.HostConfig.deployDirectory Deployment of web application directory [/var/www/ROOT/database] has finished in [8] ms
20-Mar-2019 08:19:34.641 INFO [main] org.apache.coyote.AbstractProtocol.start Starting ProtocolHandler ["http-nio-8090"]
20-Mar-2019 08:19:34.652 INFO [main] org.apache.coyote.AbstractProtocol.start Starting ProtocolHandler ["ajp-nio-8009"]
20-Mar-2019 08:19:34.654 INFO [main] org.apache.catalina.startup.Catalina.start Server startup in 19598 ms
20-Mar-2019 08:27:17.675 INFO [Thread-10] org.apache.coyote.AbstractProtocol.pause Pausing ProtocolHandler ["http-nio-8090"]
20-Mar-2019 08:27:17.746 INFO [Thread-10] org.apache.coyote.AbstractProtocol.pause Pausing ProtocolHandler ["ajp-nio-8009"]
20-Mar-2019 08:27:17.797 INFO [Thread-10] org.apache.catalina.core.StandardService.stopInternal Stopping service [Catalina]
20-Mar-2019 08:27:20.435 WARNING [localhost-startStop-2] org.apache.catalina.loader.WebappClassLoaderBase.clearReferencesThreads The web application [ROOT] appears to have started a thread named [UniqueName0.011161357303395625_Worker-1] but has failed to stop it. This is very likely to create a memory leak. Stack trace of thread:
 java.lang.Object.wait(Native Method)
 org.quartz.simpl.SimpleThreadPool$WorkerThread.run(SimpleThreadPool.java:568)
20-Mar-2019 08:27:20.509 INFO [Thread-10] org.apache.coyote.AbstractProtocol.stop Stopping ProtocolHandler ["http-nio-8090"]
20-Mar-2019 08:27:20.511 INFO [Thread-10] org.apache.coyote.AbstractProtocol.stop Stopping ProtocolHandler ["ajp-nio-8009"]
20-Mar-2019 08:27:21.943 INFO [Thread-10] org.apache.coyote.AbstractProtocol.destroy Destroying ProtocolHandler ["http-nio-8090"]
20-Mar-2019 08:27:21.944 INFO [Thread-10] org.apache.coyote.AbstractProtocol.destroy Destroying ProtocolHandler ["ajp-nio-8009"]
```

**Comment by rubendel** [comment]
This specific API call currently loads the whole model from disk in order to determine the types used. The underlying database will cache all this data, which explains the huge memory spike. I don't think this is a memory leak. To detect a memory leak, just running a method once is not really sufficient, especially not when caching is involved.

Fortunately, the types used in a model are actually also stored as metadata these days (for other optimizations), it will simply be a matter of updating this API call to actually use the meta data. It will be much faster and use almost no memory. I'll keep this issue open to implement this. I will try to get it into 1.5.126.

**Comment by motjokk** [comment]
That is good news, and thank you for the quick response. I have tried this over several days, with different files. The most interesting case is this IFC4 (37,7 kb), witch make the server stop responding completely.

The ifc-file:
[IFC4-testfil.zip](https://github.com/opensourceBIM/BIMserver/files/2987780/IFC4-testfil.zip)

Looking forward to 1.5.126 :+1:

**Comment by rubendel** [comment]
Fixed in git master, will be in 1.5.126. Unit test added as well.

Also tested with your example file (thank you), results in <10ms:
```json
{
  "response": {
    "result": [
      "IfcActorRole",
      "IfcApplication",
      "IfcAxis2Placement3D",
      "IfcBuilding",
      "IfcBuildingStorey",
      "IfcCartesianPoint",
      "IfcCartesianPointList3D",
      "IfcColourRgb",
      "IfcConversionBasedUnit",
      "IfcDerivedUnit",
      "IfcDerivedUnitElement",
      "IfcDimensionalExponents",
      "IfcDirection",
      "IfcGeometricCurveSet",
      "IfcGeometricRepresentationContext",
      "IfcGeometricRepresentationSubContext",
      "IfcLocalPlacement",
      "IfcMaterial",
      "IfcMaterialDefinitionRepresentation",
      "IfcMeasureWithUnit",
      "IfcMonetaryUnit",
      "IfcOrganization",
      "IfcOwnerHistory",
      "IfcPerson",
      "IfcPersonAndOrganization",
      "IfcPolyline",
      "IfcPostalAddress",
      "IfcPresentationLayerAssignment",
      "IfcPresentationStyleAssignment",
      "IfcProductDefinitionShape",
      "IfcProject",
      "IfcPropertyEnumeratedValue",
      "IfcPropertyEnumeration",
      "IfcPropertySet",
      "IfcPropertySingleValue",
      "IfcRelAggregates",
      "IfcRelAssociatesMaterial",
      "IfcRelConnectsPathElements",
      "IfcRelContainedInSpatialStructure",
      "IfcRelDefinesByProperties",
      "IfcRelDefinesByType",
      "IfcSIUnit",
      "IfcShapeAspect",
      "IfcShapeRepresentation",
      "IfcSite",
      "IfcSpace",
      "IfcSpaceType",
      "IfcStyledItem",
      "IfcStyledRepresentation",
      "IfcSurfaceStyle",
      "IfcSurfaceStyleRendering",
      "IfcTelecomAddress",
      "IfcTriangulatedFaceSet",
      "IfcUnitAssignment",
      "IfcWall",
      "IfcWallType"
    ]
  }
}
```
